packages/client/src/generation/TSClient/PrismaClient.ts

```diff
@@ -102,52 +151,42 @@ function clientExtensionsClientDefinition(this: PrismaClientClass) {
   }
 }
 
-function clientExtensionsDefinition(this: PrismaClientClass) {
+function clientExtensionsDefinitions(this: PrismaClientClass) {
   const result = clientExtensionsResultDefinition.call(this)
   const model = clientExtensionsModelDefinition.call(this)
   const client = clientExtensionsClientDefinition.call(this)
   const query = clientExtensionsQueryDefinition.call(this)
   const lcModelNames = Object.keys(this.dmmf.getModelMap()).map(lowerCase)
+  const modelNameUnion = lcModelNames.map((m) => `'${m}'`).join(' | ')
 
-  const definition = () => `
-  /**
-   * Allows you to extend the Prisma Client with custom logic.
-   * 
-   * Read more in our [docs](https://www.prisma.io/docs/concepts/components/prisma-client/extends).
-   */
-  $extends<
-    ${result.genericParams},
-    ${model.genericParams},
-    ${query.genericParams},
-    ${client.genericParams},
-  >(extension: {
-    result?: R & ${result.params}
-    model?: M & ${model.params}
-    query?: ${query.params}
-    client?: C & ${client.params}
-  }): runtime.Types.Utils.Omit<PrismaClient<T, U, GlobalReject, {
-        result: {${lcModelNames.reduce((acc, modelName) => {
-          return `${acc}
-          ${modelName}: ${Patch(
-            Patch(`(R & {})['$allModels']`, `(R & {})['${modelName}']`),
-            `(ExtArgs['result'] & {})['${modelName}']`,
-          )}`
-        }, '')}
-        }
-        model: {${lcModelNames.reduce((acc, modelName) => {
-          return `${acc}
-          ${modelName}: ${Patch(
-            Patch(`(M & {})['$allModels']`, `(M & {})['${modelName}']`),
-            `(ExtArgs['model'] & {})['${modelName}']`,
-          )}`
-        }, '')}
-        }
-        client: ${Patch(Pick(`C`, '`$${string}`'), `ExtArgs['client']`)}
-        query: {},
-      }>, keyof C & \`\$\${string}\`> & ${Patch(Pick(`C`, '`$${string}`'), `ExtArgs['client']`)}
-`
-
-  return ifExtensions(definition, '')
+  return {
+    prismaNamespaceDefinitions: ifExtensions(query.prismaNamespaceDefinitions, ''),
+    prismaClientDefinitions: ifExtensions(
+      `
+    /**
+     * Allows you to extend the Prisma Client with custom logic.
+     * 
+     * Read more in our [docs](https://www.prisma.io/docs/concepts/components/prisma-client/extends).
+     */
+    $extends<
+      ${result.genericParams},
+      ${model.genericParams},
+      ${query.genericParams},
+      ${client.genericParams},
+    >(extension: {
+      result?: R & ${result.params}
+      model?: M & ${model.params}
+      query?: ${query.params}
+      client?: C & ${client.params}
+    }): runtime.Types.Utils.PatchFlat3<C, PrismaClient<T, U, GlobalReject, {
+          result: { [K in ${modelNameUnion}]: runtime.Types.Utils.PatchFlat3<R[K], R['$allModels'], ExtArgs['result'][K]> } 
```

Comment:
Greatly improved performance even on big schemas, and simplified merging logic via `PatchFlat3` (patches 3 objects on top of each other), which allows us to cut on iteration/looping in the compiler.

Reply:
```ts
runtime.Types.Utils.PatchFlat3<R[K], R['$allModels'], ExtArgs['result'][K]>
```
What this means in English:
* `R[K]`: specific model extension in current input
* `R['$allModels']`: $allModels extension in current input
* `ExtArgs['result'][K]`: all previously merged models

If either of these is empty (unknown), they count as an empty object and get skipped. In sum, we can summarize `PatchFlat3<A, B, C>` as: "`A` gets applied extra properties of `B` which get applies extra properties of `C`".